doc/source/development/testing.rst

```diff
@@ -121,3 +121,164 @@ follows:
    reference the suppression files and other common arguments in a
    ``~/.valgrindrc`` file.
 
+
+Recommendations on how to write new tests
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Python-based tests should be preferred when possible, as productivity is higher
+in Python and there is no associated compilation time (compilation time affects
+feedback received from continuous integration).
+
+C/C++-based test should be reserved for C++-specific aspects that cannot be tested
+with the SWIG Python bindings, which use the C interface. For example testing
+of C++ operators (copy/move constructors/assignment operators, iterator interfaces,
+etc.) or C/C++ functionality not mapped to SWIG (e.g. CPL utility functions/classes)
+
+Python tests
+------------
+
+Python tests use the `pytest <https://docs.pytest.org/en/latest/>`__
+framework since :ref:`rfc-72`.
+
+Test cases should be written in a way where they are independent from other
+ones, so they can potentially be run in a isolated way or in parallel of other
+test cases. In particular temporary files should be created with a name that
+cannot conflict with other tests.
+
+Use ``@pytest.mark.require_driver(driver_name)`` as an annotation for a test
+case that requires an optional driver to be present.
+
+Use ``pytestmark = pytest.mark.require_driver("driver_name")`` towards the
+beginning of a test file that requires a given driver to be available for
+all its test cases. This is typically when writing tests for a particular
+driver.
+
+Use ``@pytest.mark.require_run_on_demand`` as an annotation to signal a test
+that should not be run by default, typically because it requires special
+pre-conditions, use a lot of RAM, etc. and is thus not appropriate to be
+automatically run by continuous integration.
+
+Use ``@pytest.mark.parametrize(...)`` as an annotation for test functions
+that test for variations, instead of for() constructs.
+More details at https://docs.pytest.org/en/latest/parametrize.html
+
+e.g.:
+
+.. code-block:: python
+
+    @pytest.mark.parametrize("dt,expected_size", [(gdal.GDT_Byte, 1),
+                                                  (gdal.GDT_UInt16, 2)]
+    def test_datatypesize(dt,expected_size):
+        assert gdal.GetDataTypeSizeBytes(dt) == expected_size
```

Comment:
```suggestion
.. code-block:: python

    @pytest.mark.parametrize("dt,expected_size", [(gdal.GDT_Byte, 1),
                                                  (gdal.GDT_UInt16, 2)]
    def test_datatypesize(dt,expected_size):
        assert gdal.GetDataTypeSizeBytes(dt) == expected_size

instead of

.. code-block:: python

    def test_datatypesize():
        for dt, expected_size in [(gdal.GDT_Byte, 1), (gdal.GDT_UInt16, 2)]:
            assert gdal.GetDataTypeSizeBytes(dt) == expected_size
```